python_configure.bzl

```diff
@@ -441,3 +443,62 @@ Args:
       will configure for that Python version instead of the installed
       binary. This configuration takes precedence over python_version.
 """
+
+def _parse_my_own_version_from_module_dot_bazel(ctx):
+    lines = ctx.read(Label("//:MODULE.bazel")).split("\n")
+    for line in lines:
+        parts = line.split("\"")
+        if parts[0] == "    version = ":
+            return parts[1]
+    _fail("Failed to parse my own version from `MODULE.bazel`! " +
+          "This should never happen!")
+
+def _extension_impl(ctx):
+    toolchain = None
+    for module in ctx.modules:
+        if module.is_root:
+            if not module.tags.toolchain:
+                pass
+            elif len(module.tags.toolchain) == 1:
+                toolchain = module.tags.toolchain[0]
+            else:
+                _fail("The root module may not specify multiple `toolchain` " +
+                      "tags. Found %r `toolchain` tags specified." % (
+                          len(module.tags.toolchain),
+                      ))
+        elif module.tags.toolchain:
+            _fail("A non-root module may not specify any `toolchain` tags. " +
+                  "Found %r `toolchain` tags specified by module %r." % (
+                      len(module.tags.toolchain),
+                      module.name,
+                  ))
+    if toolchain == None:
```

Comment:
@rickeylev Do we still recommend people to have `local_config_python`? I guess the toolchains from rules_python don't yet support python binding? Should they?

Reply:
It would be nice to use `@rules_python` eventually, but pybind11_bazel currently supports [embedding](https://github.com/pybind/pybind11_bazel/blob/ff261d2e9190955d0830040b20ea59ab9dbe66c8/README.md?plain=1#L19-L31): `@pybind11` has a [`pybind11_embed`](https://github.com/pybind/pybind11_bazel/blob/ff261d2e9190955d0830040b20ea59ab9dbe66c8/pybind11.BUILD#L41-L50) target for which `@local_config_python` has a [`python_embed`](https://github.com/pybind/pybind11_bazel/blob/ff261d2e9190955d0830040b20ea59ab9dbe66c8/py/BUILD.tpl#L55-L65) target; generating the latter requires interrogating the Python binary in order to determine `copts` and `linkopts`, so more just than `includes` are involved.

Leaving this open to give @rickeylev (and @rwgk) a chance to chime in. :)

Reply:
I had forgotten about `@local_config_python`'s [`python_lib`](https://github.com/pybind/pybind11_bazel/blob/ff261d2e9190955d0830040b20ea59ab9dbe66c8/py/BUILD.tpl#L33-L43) target, but now that I'm recalling this particular Windows wrinkle, I'm remembering @rickeylev's recent efforts for `@rules_python`. [#soon](https://tenor.com/P9PU.gif)

Reply:
Sorry, I lack context and this code isn't familiar to me.

What does "embedding" mean in this context? Does it mean embedding the entire Python runtime itself?

Also: Use `@rules_python//python/cc:current_py_cc_headers` to get the Python headers for the current toolchain. Really, that should be the only way you use to get the headers; other ways risk skew between build-time selection and what get used at runtime.

If you want to define your own toolchain (which is basically what local_config_python should be), then you can manually define a py_cc_toolchain() target to carry what you need.

I haven't tried it with a local system configured Python, so there might be some rough edges there; I don't see why it wouldn't work, though (the main input is just a cc_library)

> more just than includes are involved.

Like what? Can you post to https://github.com/bazelbuild/rules_python/issues/824 about what, besides headers, is needed? I was wondering exactly this myself when I implemented py_cc_toolchain. It should be fairly easy to expand what the toolchain carries (I wrote it with that in mind)

(I also just saw that ethanluoyc posted some recent comments about using pybind11 with py_cc_toolchain which sound promising)

Reply:
> What does "embedding" mean in this context? Does it mean embedding the entire Python runtime itself?

See https://docs.python.org/3/extending/embedding.html, I guess, but the TLDR is that "extending" enables Python to call out to C/C++ whereas "embedding" enables C/C++ to call out to Python. pybind11_bazel currently facilitates both, so I need to preserve that. In fact, as much as I would like to use `@rules_python//python/cc:current_py_cc_headers`, I need to make pybind11_bazel **_as is_** support Bzlmod – and add it to BCR! – before exploring `@rules_python` adoption.

> Can you post to bazelbuild/rules_python#824 about what, besides headers, is needed?

Sure, I will follow up there. :)